File: backend/app/api/v1/endpoints/campaigns.py
# =============================================================================
# Stratum AI - Campaign Management Endpoints
# =============================================================================
"""
Campaign CRUD operations and metrics retrieval.
Implements Module B: Unified Campaign Model.
"""

from datetime import date, datetime, timedelta, timezone
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, Request, status
from sqlalchemy import and_, func, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import load_only

from app.auth.deps import get_current_user
from app.core.logging import get_logger
from app.db.session import get_async_session
from app.models import AdPlatform, Campaign, CampaignMetric, CampaignStatus
from app.schemas import (
    APIResponse,
    CampaignCreate,
    CampaignDetailResponse,
    CampaignListResponse,
    CampaignMetricResponse,
    CampaignMetricsTimeSeriesResponse,
    CampaignResponse,
    CampaignUpdate,
    PaginatedResponse,
)

logger = get_logger(__name__)
router = APIRouter(
    # SECURITY (STRAT-SC-001/C3): the old per-org guards this router relied
    # on were deleted in the de-tenanting sweep; real auth now enforced here.
    dependencies=[Depends(get_current_user)],
)


@router.get("", response_model=APIResponse[PaginatedResponse[CampaignListResponse]])
async def list_campaigns(
    db: AsyncSession = Depends(get_async_session),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    platform: Optional[AdPlatform] = None,
    status: Optional[CampaignStatus] = None,
    search: Optional[str] = None,
    labels: Optional[List[str]] = Query(None),
    account_id: Optional[str] = Query(None, description="Filter by ad account ID"),
):
    """
    List campaigns with filtering and pagination.

    Args:
        page: Page number (1-indexed)
        page_size: Number of items per page
        platform: Filter by ad platform
        status: Filter by campaign status
        search: Search by name
        labels: Filter by labels
    """

    # Use only indexed columns for WHERE
    query = select(Campaign).where(
        Campaign.is_deleted == False,
    )

    # Apply filters
    if platform:
        query = query.where(Campaign.platform == platform)
    if status:
        query = query.where(Campaign.status == status)
    if search:
        query = query.where(Campaign.name.ilike(f"%{search}%"))
    if account_id:
        query = query.where(Campaign.account_id == account_id)
    if labels:
        # PostgreSQL JSONB contains any of the labels
        for label in labels:
            query = query.where(Campaign.labels.contains([label]))

    # Count total — use COUNT(*) directly without loading full rows
    count_query = select(func.count()).select_from(query.subquery())
    total_result = await db.execute(count_query)
    total = total_result.scalar()

    # Pagination — defer loading heavy JSONB columns not needed for list view
    offset = (page - 1) * page_size
    query = (
        query.order_by(Campaign.updated_at.desc())
        .offset(offset)
        .limit(page_size)
        .options(
            load_only(
                Campaign.id,
                Campaign.name,
                Campaign.platform,
                Campaign.status,
                Campaign.total_spend_cents,
                Campaign.impressions,
                Campaign.clicks,
                Campaign.conversions,
                Campaign.roas,
                Campaign.labels,
                Campaign.last_synced_at,
            )
        )
    )

    result = await db.execute(query)
    campaigns = result.scalars().all()

    return APIResponse(
        success=True,
        data=PaginatedResponse(
            items=[
                CampaignListResponse(
                    id=c.id,
                    name=c.name,
                    platform=c.platform,
                    status=c.status,
                    total_spend_cents=c.total_spend_cents,
                    impressions=c.impressions,
                    clicks=c.clicks,
                    conversions=c.conversions,
                    roas=c.roas,
                    labels=c.labels,
                    last_synced_at=c.last_synced_at,
                )
                for c in campaigns
            ],
            total=total,
            page=page,
            page_size=page_size,
            total_pages=(total + page_size - 1) // page_size,
        ),
    )


@router.get("/{campaign_id}", response_model=APIResponse[CampaignDetailResponse])
async def get_campaign(
    campaign_id: int,
    db: AsyncSession = Depends(get_async_session),
):
    """Get detailed campaign information."""

    result = await db.execute(
        select(Campaign).where(
            Campaign.id == campaign_id,
            Campaign.is_deleted == False,
        )
    )
    campaign = result.scalar_one_or_none()

    if not campaign:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Campaign not found",
        )

    return APIResponse(
        success=True,
        data=CampaignDetailResponse(
            id=campaign.id,
            name=campaign.name,
            platform=campaign.platform,
            external_id=campaign.external_id,
            account_id=campaign.account_id,
            status=campaign.status,
            objective=campaign.objective,
            daily_budget_cents=campaign.daily_budget_cents,
            lifetime_budget_cents=campaign.lifetime_budget_cents,
            total_spend_cents=campaign.total_spend_cents,
            currency=campaign.currency,
            impressions=campaign.impressions,
            clicks=campaign.clicks,
            conversions=campaign.conversions,
            revenue_cents=campaign.revenue_cents,
            ctr=campaign.ctr,
            cpc_cents=campaign.cpc_cents,
            cpm_cents=campaign.cpm_cents,
            cpa_cents=campaign.cpa_cents,
            roas=campaign.roas,
            targeting_age_min=campaign.targeting_age_min,
            targeting_age_max=campaign.targeting_age_max,
            targeting_genders=campaign.targeting_genders,
            targeting_locations=campaign.targeting_locations,
            start_date=campaign.start_date,
            end_date=campaign.end_date,
            labels=campaign.labels,
            last_synced_at=campaign.last_synced_at,
            demographics_age=campaign.demographics_age,
            demographics_gender=campaign.demographics_gender,
            demographics_location=campaign.demographics_location,
            created_at=campaign.created_at,
            updated_at=campaign.updated_at,
        ),
    )


@router.post(
    "",
    response_model=APIResponse[CampaignResponse],
    status_code=status.HTTP_201_CREATED,
)
async def create_campaign(
    campaign_data: CampaignCreate,
    db: AsyncSession = Depends(get_async_session),
):
    """Create a new campaign."""

    # Check for duplicate
    existing = await db.execute(
        select(Campaign).where(
            Campaign.platform == campaign_data.platform,
            Campaign.external_id == campaign_data.external_id,
        )
    )
    if existing.scalar_one_or_none():
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Campaign with this external_id already exists",
        )

    campaign = Campaign(
        **campaign_data.model_dump(),
    )

    db.add(campaign)
    await db.commit()

    logger.info("campaign_created", campaign_id=campaign.id)

    return APIResponse(
        success=True,
        data=CampaignResponse.model_validate(campaign),
        message="Campaign created successfully",
    )


@router.patch("/{campaign_id}", response_model=APIResponse[CampaignResponse])
async def update_campaign(
    campaign_id: int,
    update_data: CampaignUpdate,
    db: AsyncSession = Depends(get_async_session),
):
    """Update a campaign."""

    result = await db.execute(
        select(Campaign).where(
            Campaign.id == campaign_id,
            Campaign.is_deleted == False,
        )
    )
    campaign = result.scalar_one_or_none()

    if not campaign:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Campaign not found",
        )

    # Update fields
    for field, value in update_data.model_dump(exclude_unset=True).items():
        setattr(campaign, field, value)

    await db.commit()

    logger.info("campaign_updated", campaign_id=campaign_id)

    return APIResponse(
        success=True,
        data=CampaignResponse.model_validate(campaign),
        message="Campaign updated successfully",
    )


@router.delete("/{campaign_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_campaign(
    campaign_id: int,
    db: AsyncSession = Depends(get_async_session),
):
    """Soft delete a campaign."""

    result = await db.execute(
        select(Campaign).where(
            Campaign.id == campaign_id,
            Campaign.is_deleted == False,
        )
    )
    campaign = result.scalar_one_or_none()

    if not campaign:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Campaign not found",
        )

    campaign.soft_delete()
    await db.commit()

    logger.info("campaign_deleted", campaign_id=campaign_id)


@router.get(
    "/{campaign_id}/metrics",
    response_model=APIResponse[CampaignMetricsTimeSeriesResponse],
)
async def get_campaign_metrics(
    campaign_id: int,
    db: AsyncSession = Depends(get_async_session),
    start_date: Optional[date] = None,
    end_date: Optional[date] = None,
):
    """
    Get time-series metrics for a campaign.

    Args:
        campaign_id: Campaign ID
        start_date: Start date (default: 30 days ago)
        end_date: End date (default: today)
    """

    # Verify campaign exists
    campaign_result = await db.execute(
        select(Campaign).where(
            Campaign.id == campaign_id,
        )
    )
    if not campaign_result.scalar_one_or_none():
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Campaign not found",
        )

    # Default date range
    if not end_date:
        end_date = date.today()
    if not start_date:
        start_date = end_date - timedelta(days=30)

    # Query metrics
    result = await db.execute(
        select(CampaignMetric)
        .where(
            CampaignMetric.campaign_id == campaign_id,
            CampaignMetric.date >= start_date,
            CampaignMetric.date <= end_date,
        )
        .order_by(CampaignMetric.date)
        .limit(1000)
    )
    metrics = result.scalars().all()

    # Calculate aggregates
    total_impressions = sum(m.impressions for m in metrics)
    total_clicks = sum(m.clicks for m in metrics)
    total_conversions = sum(m.conversions for m in metrics)
    total_spend = sum(m.spend_cents for m in metrics)
    total_revenue = sum(m.revenue_cents for m in metrics)

    aggregated = {
        "impressions": total_impressions,
        "clicks": total_clicks,
        "conversions": total_conversions,
        "spend_cents": total_spend,
        "revenue_cents": total_revenue,
        "ctr": (total_clicks / total_impressions * 100) if total_impressions > 0 else 0,
        "roas": (total_revenue / total_spend) if total_spend > 0 else 0,
    }

    return APIResponse(
        success=True,
        data=CampaignMetricsTimeSeriesResponse(
            campaign_id=campaign_id,
            date_range={"start": start_date.isoformat(), "end": end_date.isoformat()},
            metrics=[
                CampaignMetricResponse(
                    date=m.date,
                    impressions=m.impressions,
                    clicks=m.clicks,
                    conversions=m.conversions,
                    spend_cents=m.spend_cents,
                    revenue_cents=m.revenue_cents,
                    video_views=m.video_views,
                    video_completions=m.video_completions,
                )
                for m in metrics
            ],
            aggregated=aggregated,
        ),
    )


@router.post("/sync-all")
async def trigger_sync_all_campaigns(
    request: Request,
) -> APIResponse:
    """
    Trigger a manual sync for all campaigns across all platforms.
    Queues a Celery task to fetch latest data from every ad platform.
    """
    from app.workers.tasks import sync_all_campaigns

    task = sync_all_campaigns.delay()

    return APIResponse(
        success=True,
        data={"task_id": task.id},
        message="Sync all campaigns queued successfully",
    )


@router.post("/sync-platform/{platform}")
async def trigger_platform_sync(
    platform: str,
    db: AsyncSession = Depends(get_async_session),
) -> APIResponse:
    """
    Discover and sync all campaigns from a specific ad platform.

    This calls the orchestrator to fetch campaigns directly from the
    platform API (Meta, TikTok, Snapchat) and upsert them into the DB.
    Use this to force-discover new campaigns that haven't been seen before.

    NOTE(C3): this router previously had no per-request auth dependency on
    any of its 8 routes — it relied solely on the removed tenant-scoping
    middleware to reject requests (a check that would otherwise always 401,
    since that middleware no longer sets any such context). CLOSED in the
    post-review fix loop: `router = APIRouter(dependencies=[Depends(get_current_user)])`
    now gates every route in this file, matching the pattern applied to the
    11 sibling routers (rules.py, qa_fixes.py, etc.).
    """
    from app.services.sync.orchestrator import PlatformSyncOrchestrator

    try:
        ad_platform = AdPlatform(platform.lower())
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Unknown platform: {platform}. Use: meta, tiktok, snapchat, google",
        )

    orchestrator = PlatformSyncOrchestrator(db)
    result = await orchestrator.sync_platform(ad_platform, days_back=30)

    return APIResponse(
        success=len(result.errors) == 0,
        data={
            "platform": result.platform,
            "campaigns_synced": result.campaigns_synced,
            "metrics_upserted": result.metrics_upserted,
            "duration_seconds": result.duration_seconds,
            "errors": result.errors,
        },
        message=(
            f"Platform sync completed: {result.campaigns_synced} campaigns synced"
            if not result.errors
            else f"Platform sync finished with errors: {', '.join(result.errors)}"
        ),
    )


@router.post("/{campaign_id}/sync")
async def trigger_campaign_sync(
    campaign_id: int,
    db: AsyncSession = Depends(get_async_session),
):
    """
    Trigger a manual sync for a campaign.
    Queues a Celery task to fetch latest data from the ad platform.
    """

    # Verify campaign exists
    result = await db.execute(
        select(Campaign).where(
            Campaign.id == campaign_id,
        )
    )
    campaign = result.scalar_one_or_none()

    if not campaign:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Campaign not found",
        )

    # Queue sync task
    from app.workers.tasks import sync_campaign_data

    task = sync_campaign_data.delay(campaign_id)

    return APIResponse(
        success=True,
        data={"task_id": task.id},
        message="Sync queued successfully",
    )

File: frontend/src/api/campaigns.ts
/**
 * ADs Growth System - Campaigns API
 *
 * Campaign management endpoints for CRUD operations and metrics
 */

import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query'
import { apiClient, ApiResponse, PaginatedResponse } from './client'

// Types
export type Platform = 'meta' | 'google' | 'tiktok' | 'snapchat'
export type CampaignStatus = 'active' | 'paused' | 'completed' | 'draft' | 'error'
export type CampaignObjective = 'awareness' | 'consideration' | 'conversion'

export interface Campaign {
  id: string
  tenantId: number
  platform: Platform
  platformCampaignId: string
  name: string
  status: CampaignStatus
  objective: CampaignObjective
  budget: number
  dailyBudget: number | null
  currency: string
  startDate: string
  endDate: string | null
  createdAt: string
  updatedAt: string
  lastSyncedAt: string | null
}

export interface CampaignMetrics {
  campaignId: string
  date: string
  spend: number
  impressions: number
  clicks: number
  conversions: number
  revenue: number
  roas: number
  ctr: number
  cpc: number
  cpa: number
  reach: number | null
  frequency: number | null
}

export interface CampaignWithMetrics extends Campaign {
  metrics: CampaignMetrics
  previousMetrics?: CampaignMetrics
}

export interface CampaignFilters {
  platform?: Platform
  status?: CampaignStatus
  objective?: CampaignObjective
  startDate?: string
  endDate?: string
  search?: string
  skip?: number
  limit?: number
}

export interface CreateCampaignRequest {
  platform: Platform
  name: string
  objective: CampaignObjective
  budget: number
  dailyBudget?: number
  currency?: string
  startDate: string
  endDate?: string
  targeting?: Record<string, unknown>
  creatives?: string[]
}

export interface UpdateCampaignRequest {
  name?: string
  status?: CampaignStatus
  budget?: number
  dailyBudget?: number
  endDate?: string
  targeting?: Record<string, unknown>
}

// API Functions
export const campaignsApi = {
  /**
   * Get all campaigns with optional filters
   */
  getCampaigns: async (filters: CampaignFilters = {}): Promise<PaginatedResponse<CampaignWithMetrics>> => {
    const safeFilters = { limit: 50, skip: 0, ...filters }
    const response = await apiClient.get<ApiResponse<PaginatedResponse<CampaignWithMetrics>>>(
      '/campaigns',
      { params: safeFilters }
    )
    return response.data.data
  },

  /**
   * Get a single campaign by ID
   */
  getCampaign: async (id: string): Promise<CampaignWithMetrics> => {
    const response = await apiClient.get<ApiResponse<CampaignWithMetrics>>(`/campaigns/${id}`)
    return response.data.data
  },

  /**
   * Create a new campaign
   */
  createCampaign: async (data: CreateCampaignRequest): Promise<Campaign> => {
    const response = await apiClient.post<ApiResponse<Campaign>>('/campaigns', data)
    return response.data.data
  },

  /**
   * Update an existing campaign
   */
  updateCampaign: async (id: string, data: UpdateCampaignRequest): Promise<Campaign> => {
    const response = await apiClient.patch<ApiResponse<Campaign>>(`/campaigns/${id}`, data)
    return response.data.data
  },

  /**
   * Delete a campaign
   */
  deleteCampaign: async (id: string): Promise<void> => {
    await apiClient.delete(`/campaigns/${id}`)
  },

  /**
   * Get campaign metrics
   */
  getCampaignMetrics: async (
    id: string,
    startDate: string,
    endDate: string,
    granularity: 'day' | 'week' | 'month' = 'day'
  ): Promise<CampaignMetrics[]> => {
    const response = await apiClient.get<ApiResponse<CampaignMetrics[]>>(
      `/campaigns/${id}/metrics`,
      { params: { start_date: startDate, end_date: endDate, granularity } }
    )
    return response.data.data
  },

  /**
   * Sync campaign from platform
   */
  syncCampaign: async (id: string): Promise<Campaign> => {
    const response = await apiClient.post<ApiResponse<Campaign>>(`/campaigns/${id}/sync`)
    return response.data.data
  },

  /**
   * Sync all campaigns from all platforms
   */
  syncAllCampaigns: async (): Promise<{ task_id: string; message: string }> => {
    const response = await apiClient.post<ApiResponse<{ task_id: string; message: string }>>('/campaigns/sync-all')
    return response.data.data
  },

  /**
   * Bulk update campaign status
   */
  bulkUpdateStatus: async (
    ids: string[],
    status: CampaignStatus
  ): Promise<Campaign[]> => {
    const response = await apiClient.post<ApiResponse<Campaign[]>>(
      '/campaigns/bulk/status',
      { ids, status }
    )
    return response.data.data
  },

  /**
   * Pause a campaign
   */
  pauseCampaign: async (id: string): Promise<Campaign> => {
    const response = await apiClient.post<ApiResponse<Campaign>>(`/campaigns/${id}/pause`)
    return response.data.data
  },

  /**
   * Activate a campaign
   */
  activateCampaign: async (id: string): Promise<Campaign> => {
    const response = await apiClient.post<ApiResponse<Campaign>>(`/campaigns/${id}/activate`)
    return response.data.data
  },
}

// React Query Hooks

export function useCampaigns(filters: CampaignFilters = {}) {
  return useQuery({
    queryKey: ['campaigns', filters],
    queryFn: () => campaignsApi.getCampaigns(filters),
    staleTime: 30 * 1000,
  })
}

export function useCampaign(id: string) {
  return useQuery({
    queryKey: ['campaigns', id],
    queryFn: () => campaignsApi.getCampaign(id),
    enabled: !!id,
  })
}

export function useCreateCampaign() {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: campaignsApi.createCampaign,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['campaigns'] })
    },
  })
}

export function useUpdateCampaign() {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: ({ id, data }: { id: string; data: UpdateCampaignRequest }) =>
      campaignsApi.updateCampaign(id, data),
    onSuccess: (_, variables) => {
      queryClient.invalidateQueries({ queryKey: ['campaigns'] })
      queryClient.invalidateQueries({ queryKey: ['campaigns', variables.id] })
    },
  })
}

export function useDeleteCampaign() {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: campaignsApi.deleteCampaign,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['campaigns'] })
    },
  })
}

export function useCampaignMetrics(
  id: string,
  startDate: string,
  endDate: string,
  granularity: 'day' | 'week' | 'month' = 'day'
) {
  return useQuery({
    queryKey: ['campaigns', id, 'metrics', startDate, endDate, granularity],
    queryFn: () => campaignsApi.getCampaignMetrics(id, startDate, endDate, granularity),
    enabled: !!id && !!startDate && !!endDate,
  })
}

export function useSyncCampaign() {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: campaignsApi.syncCampaign,
    onSuccess: (_, id) => {
      queryClient.invalidateQueries({ queryKey: ['campaigns', id] })
    },
  })
}

export function useSyncAllCampaigns() {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: campaignsApi.syncAllCampaigns,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['campaigns'] })
    },
  })
}

export function useBulkUpdateCampaignStatus() {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: ({ ids, status }: { ids: string[]; status: CampaignStatus }) =>
      campaignsApi.bulkUpdateStatus(ids, status),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['campaigns'] })
    },
  })
}

export function usePauseCampaign() {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: campaignsApi.pauseCampaign,
    onSuccess: (_, id) => {
      queryClient.invalidateQueries({ queryKey: ['campaigns'] })
      queryClient.invalidateQueries({ queryKey: ['campaigns', id] })
    },
  })
}

export function useActivateCampaign() {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: campaignsApi.activateCampaign,
    onSuccess: (_, id) => {
      queryClient.invalidateQueries({ queryKey: ['campaigns'] })
      queryClient.invalidateQueries({ queryKey: ['campaigns', id] })
    },
  })
}
